### PR TITLE
JESI-5965

### PR DESCRIPTION
**Pull request by AndreTheHunter**
JESI-5965
# Checklist

## Pre-Merge

- [x] Version updated in accordance with [Semver](https://semver.org)
- [x] Public functions, variables, and objects have doc strings with `@since ...` tags
- [x] Public functions and objects have unit tests
- [x] All updated captured in `CHANGELOG.md`
- [x] CI/CD pipeline passes
- [x] `dart analyze` and `dart format --fix .` has been run

## Post Merge

- [x] Tag repository with the latest version. `git tag x.y.z`
- [x] Push tags to GitHub. `git push -u --tags`

### CODE REVIEW

File: lib/src/file_system_entity_extensions.dart [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  /// This behavior allows [deleteExisting] to be used to unconditionally delete any file
+  /// system object.
+  ///
+  /// Returns a `Future<FileSystemEntity>` that completes with [this]
+  /// [FileSystemEntity] when the deletion is done. If the [FileSystemEntity]
+  /// cannot be deleted, the future completes with an exception.
+  ///
+  /// `since 1.3.0`
+  //FIXME is this the best name?
+  Future<void> deleteExisting({
```
Comment (AndreTheHunter) [inline lib/src/file_system_entity_extensions.dart]: Ideas @axrs?

File: test/file_system_entity_extensions_test.dart [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        expect(await dir.exists(), true);
+        if (await fileInDir.exists().isFalse) {
+          await fileInDir.create();
+        }
+        expect(await fileInDir.exists(), true);
+        await dir.deleteExisting(recursive: true);
+        expect(await dir.exists(), false);
+        expect(await fileInDir.exists(), false);
+      });
+      test('does not error if the file does not exist', () async {
```
Comment (AndreTheHunter) [inline test/file_system_entity_extensions_test.dart:37]: The value add

File: test/file_system_entity_extensions_test.dart [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        if (await file.exists()) {
+          await file.delete();
+        }
+        expect(await file.exists(), false);
+        await file.deleteExisting();
+        expect(
+          () async => await file.delete(),
+          throwsA(TypeMatcher<io.FileSystemException>()),
+        );
```
Comment (AndreTheHunter) [inline test/file_system_entity_extensions_test.dart:46]: Doesn't do this (it's dumb)

File: lib/src/file_system_entity_util.dart [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  /// `since 1.3.0`
+  static Future<void> delete(
+    FileSystemEntity fileSystemEntity, {
+    bool recursive = false,
+  }) async {
+    try {
+      await fileSystemEntity.delete(recursive: recursive);
+    } on FileSystemException {
+      if (await fileSystemEntity.exists()) rethrow;
+    }
```
Comment (AndreTheHunter) [inline lib/src/file_system_entity_util.dart:37]: Moved from  https://github.com/jesims/azure-infrastructure/pull/22

File: test/file_system_entity_extensions_test.dart [hunk 5]
```diff
@@ -0,0 +1,62 @@
+import 'dart:io' as io;
+
+import 'package:file/memory.dart';
+import 'package:path/path.dart' as path;
+import 'package:pouch/pouch.dart';
+import 'package:pouch/pouch_io.dart';
+import 'package:test/test.dart';
+
+void main() {
+  var fs = MemoryFileSystem();
```
Comment (AndreTheHunter) [inline test/file_system_entity_extensions_test.dart:9]: ❤️ an abstracted file system that can work in memory!

File: lib/src/file_system_entity_extensions.dart (hunk 1 above)
Comment (axrs) [reply lib/src/file_system_entity_extensions.dart]: let's call it `forceDelete` after `rm -f`

File: test/file_system_entity_extensions_test.dart [hunk 6]
```diff
@@ -0,0 +1,62 @@
+import 'dart:io' as io;
```
Comment (axrs) [inline test/file_system_entity_extensions_test.dart]: Let's use universal_io

Add: https://pub.dev/packages/universal_io
Then replace all `import 'dart:io'` with `import 'package:universal_io/io.dart'`

**Review by axrs — CHANGES_REQUESTED** [review]

File: test/file_system_entity_extensions_test.dart (hunk 6 above)
Comment (AndreTheHunter) [reply test/file_system_entity_extensions_test.dart]: Was only used in one place, in a test, for an exception type, that's exports to the `dart:io` value....
but it's good practice to not import `dart:io` so I'll do it 🙂

**Review by axrs — APPROVED** [review]

**Pull request merged**